contracts/MuonNodeStaking.sol

```diff
@@ -0,0 +1,631 @@
+        require(
+            user.paidRewardPerToken <= paidRewardPerToken &&
+                paidRewardPerToken <= rewardPerToken(),
+            "Invalid paidRewardPerToken value."
+        );
+
+        // Verify the authenticity of the withdrawal request.
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                muonAppId,
+                reqId,
+                msg.sender,
+                user.paidReward,
+                paidRewardPerToken,
+                amount
+            )
+        );
+
+        bool verified = muonVerify(
+            reqId,
+            uint256(hash),
+            signature,
+            muonPublicKey
+        );
+        require(verified, "Invalid signature.");
+
+        uint256 maxReward = (user.balance *
+            (paidRewardPerToken - user.paidRewardPerToken)) /
+            1e18 +
+            user.pendingRewards;
+        require(amount <= maxReward, "Invalid withdrawal amount.");
+        notPaidRewards += (maxReward - amount);
+
+        require(amount <= earned(msg.sender), "Invalid withdrawal amount.");
+
+        users[msg.sender].pendingRewards = 0;
+        users[msg.sender].paidReward += amount;
+        users[msg.sender].paidRewardPerToken = paidRewardPerToken;
+        muonToken.transfer(msg.sender, amount);
+        emit RewardGot(reqId, msg.sender, amount);
+    }
+
+    /**
+     * @dev Allows stakers to request to exit from the network.
+     * Stakers can withdraw the staked amount after the exit pending period has passed.
+     */
+    function requestExit() public updateReward(msg.sender) {
+        IMuonNodeManager.Node memory node = nodeManager.stakerAddressInfo(
+            msg.sender
+        );
+        require(node.id != 0, "Node not found for the staker address.");
+
+        require(node.active, "The node is already deactivated.");
+
+        totalStaked -= users[msg.sender].balance;
+        users[msg.sender].balance = 0;
+        nodeManager.deactiveNode(node.id);
+        emit ExitRequested(msg.sender);
+    }
+
+    /**
+     * @dev Allows stakers to withdraw their staked amount after exiting the network and exit pending period has passed.
+     */
+    function withdraw() public {
+        IMuonNodeManager.Node memory node = nodeManager.stakerAddressInfo(
+            msg.sender
+        );
+        require(node.id != 0, "Node not found for the staker address.");
+
+        require(
+            !node.active &&
+                (node.endTime + exitPendingPeriod) < block.timestamp,
+            "The exit time has not been reached yet."
+        );
+
+        require(
+            !lockedStakes[msg.sender],
+            "Your stake is currently locked and cannot be withdrawn."
+        );
+
+        uint256 tokenId = users[msg.sender].tokenId;
+        require(tokenId != 0, "No staking found for the staker address.");
+
+        bondedToken.safeTransferFrom(address(this), msg.sender, tokenId);
+        users[msg.sender].tokenId = 0;
```

Comment:
Need to set user.balance = 0 if it is not already 0.

In some cases, admins can deactivate a node on the NodeManager contract directly.

Reply:
actually need to add this when balance is not 0:
```js
totalStaked -= users[msg.sender].balance;
users[msg.sender].balance = 0;
```

Reply:
This is done.
We know that `updateReward` should be called when balance is not 0. But it is just a security check and will not happen normally.